# sc_cluster
Single cell clustering experiments that use **AIDE** to reduce the dimension of cell-gene matrix and then use **RPH-KMeans** to do clustering.

- [AIDE](https://github.com/tinglabs/aide): Autoencoder-imputed distance-preserved embedding, which combines both **Multidimentional Scaling (MDS)** and **AutoEncoder (AE)** technique, aiming to preserve the distance between imputed data generated by **AE** when reducing dimension.
- [rph-kmeans](https://github.com/tinglabs/rph_kmeans): a variant of kmeans algorithm in which the initial centers is produced by point reduction process using **random projection (RP)**, which is one of the **local sensitive hashing (LSH)** technique.

## File Illustration
```
result
	dim_origin_clt_eval_paper: output of dim_origin_clt_eval.py
	dim_reduce_eval_paper: output of dim_reduce_clt_eval.py
	data_explain_paper: basic infomation of data
	embed_mem_use_paper: memory of running AIDE and PCA
src
	reader.py: read data from raw file
	preprocess.py: data preprocess
	dim_origin_clt_eval.py: data preprocess + clustering
	dim_reduce_clt_eval.py: data preprocess + dimension reduction + clustering
data (not uploaded)
	raw
		1M_neurons
			1M_neurons_filtered_gene_bc_matrices_h5.h5
		sc_brain
			l5_all.loom
		scDeepCluster
			PBMC_68k
				PBMC_68k.h5
			Shekhar_mouse_retina
				Shekhar_mouse_retina.h5
			10X_PBMC
				10X_PBMC.h5
			mouse_bladder_cell
				mouse_bladder_cell.h5
			mouse_ES_cell
				mouse_ES_cell.h5
			worm_neuron_cell
				worm_neuron_cell.h5
``` 

